iree/compiler/Utils/CustomKernelsTargetInfo.h

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_UTILS_CUSTOMKERNELTARGETINFO_H_
+#define IREE_COMPILER_UTILS_CUSTOMKERNELTARGETINFO_H_
+
+#include <stdint.h>
+
+#include <cassert>
+
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Support/LogicalResult.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+// Enumerates target ISAs that we care about.
+enum class CustomKernelTargetArch { None, Aarch64 };
+
+// Enumerates arch-specific target features that we care about.
+// We explicitly want to stick to the default enumeration values (0, 1, 2, ...,
+// no greater than 63) because this is going to be indexing a uint64 bitfield.
+// Intentionally not reusing bits across architectures to be able to catch
+// most bugs. 64 is enough across all target architectures for now.
+enum class CustomKernelTargetFeature {
+  // Indicates a preference for intrinsics over inline asm. Unlike other bits,
+  // this is generic, not tied to a particular architecture or CPU feature, and
+  // it has to be passed through some custom boolean flag or option, not as
+  // part of the target CPU features.
+  Intrinsics,
+  // Aarch64 features.
+  Aarch64Dotprod,
+};
+
+inline bool isFeatureForArch(CustomKernelTargetFeature feature,
+                             CustomKernelTargetArch arch) {
+  switch (feature) {
+    case CustomKernelTargetFeature::Intrinsics:
+      return true;
+    case CustomKernelTargetFeature::Aarch64Dotprod:
+      return arch == CustomKernelTargetArch::Aarch64;
+    default:
+      assert(false);
+      return false;
+  }
```

Comment:
FTR, `llvm_unreachable` was discussed on `#off-topic` IREE discord today.  TLDR is that given the danger if explicitly intentional UB is ever reached, it can only be the right trade-off of safety for performance for the most performance critical code. The present code isn't performance-critical. https://llvm.org/docs/CodingStandards.html#assert-liberally shouldn't be saying the things it's saying without clear warnings and framing as a thing only for rare performance-critical code.

Here is an interesting case because the set of allowed values for an enumeration is the bit-OR'ed combination of the enumeration values, which here are the default values 0 and 1, and so even as the spec allows also any bit-OR'ed combinations of those, that doesn't actually enable any additional values than these 0 and 1.  This means that the switch statemenet without the `default:` is technically equivalent to the one with it --- and the compiler actually shouldn't complain in this case if we omitted it (I haven't checked whether it does).  Still, that would be an instance of *likely unintentional UB* that the compiler likely wouldn't take advantage of too aggressively, so if in practice an ill-formed value made its way into here, even though it's UB, it would in practice compile in the "naive" way, and our `default:` case would be taken, so it's useful to have it with the assert and the return false so we have sane behavior in practice.

In theory though, in case of an ill-formed value, since that's UB, the compiler is allowed to assume that can't possibly happen and so it absolutely can compile the function in a way that a bad value causes the `switch` to jump past the end of the function or past the `ret` instruction...  This remains unlikely as long as the UB is *likely unintentional*. Using `llvm_unreachable` changes that into *likely intentional* and now we're really daring the compiler to show us its optimization prowess.

Reply:
Yeah that conversation has me convinced. I still think that leaving off the default so that the compiler can ensure this remains exhaustive is useful though.

```suggestion
  }
  // Note: *not* llvm_unreachable, which introduces intentionally
  // undefined behavior, which is not a good tradeoff in this
  // not-particularly-performance-sensitive context.
  assert(false && "invalid feature in switch statement");
  return false;
```